File: 0028.py
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        head = tail = 0
        while tail < len(haystack) - len(needle) + 1:
            if head-tail == len(needle):
                return tail
            if haystack[head] == needle[head-tail]:
                head += 1
            else:
                tail += 1
                head = tail
        return tail if head-tail == len(needle) else -1
    

File: src/0028.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int strStr(string haystack, string needle) {
        int head = 0, tail = 0;
        while(tail < haystack.size() - needle.size() + 1) {
            if(head-tail == needle.size()) {
                return tail;
            }
            if(haystack[head] == needle[head-tail]) {
                head++;
            }
            else {
                head = ++tail;
            }
        }
        return head-tail == needle.size() ? tail : -1;
    }
};
